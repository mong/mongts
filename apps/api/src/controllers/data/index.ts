export * from "./description";
export * from "./indicators";
export * from "./unitNames";
export * from "./selectionyears";
