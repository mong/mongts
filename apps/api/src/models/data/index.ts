export * from "./unitNames";
export * from "./indicators";
export * from "./description";
export * from "./selectionyears";

export interface Filter {
  unit_level?: string;
  year?: number;
  unit_name?: string[];
  register?: string;
  type?: string;
  context?: string;
}
