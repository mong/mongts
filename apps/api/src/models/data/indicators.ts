import { Knex } from "knex";
import db from "../../db";
import { Filter } from ".";

export interface Indicator {
  ind_id: string;
  unit_level: string;
  unit_name: string;
  orgnr: number;
  year: number;
  denominator: number;
  var: number;
  level: string;
  level_direction: number | null;
  dg?: number;
  include: number | null;
  min_denominator: number;
}

export const indicatorsModel = (filter?: Filter): Promise<Indicator[]> =>
  db
    .select("agg_data.*", "ind.type", "ind.include", "ind.min_denominator")
    .from("agg_data")
    .leftJoin("ind", "agg_data.ind_id", "ind.id")
    .where("include", 1)
    .where(function () {
      this.whereRaw("denominator >= min_denominator")
        .orWhereNull("min_denominator")
        .orWhereRaw("type IN ('dg_andel', 'dg_beregnet_andel')");
    })
    .whereNot("unit_name", "LIKE", "Udefinerte%")
    .modify(withFilter, filter);

function withFilter(builder: Knex.QueryBuilder, filter?: Filter) {
  if (filter?.unit_level) {
    builder.andWhere("unit_level", filter.unit_level);
  }
  if (filter?.year && typeof filter?.year === "number") {
    builder.where("year", filter.year);
  }
  if (filter?.unit_name) {
    builder.whereIn("unit_name", filter.unit_name);
  }
  if (filter?.context) {
    builder.where("context", filter.context);
  }
  if (filter?.type) {
    if (filter.type === "dg") {
      builder.whereIn("type", ["dg", "dg_andel", "dg_beregnet_andel"]);
    }
    if (filter.type === "ind") {
      builder.whereNotIn("type", ["dg", "dg_andel", "dg_beregnet_andel"]);
    }
  }
  if (filter?.register) {
    builder.whereIn("ind_id", function (this: Knex.QueryBuilder) {
      this.select("ind.id")
        .from("ind")
        .modify(registerFilter, filter.register ?? "");
    });
  }
}

function registerFilter(builder: Knex.QueryBuilder, registerName: string) {
  builder.where("registry_id", function (this: Knex.QueryBuilder) {
    this.select("id").from("registry").where("name", registerName);
  });
}
