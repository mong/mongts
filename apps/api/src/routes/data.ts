import express from "express";
import {
  descriptionController,
  indicatorsContoller,
  unitNamesContoller,
  selectionYearsContoller,
} from "../controllers/data";

const Router = express.Router();

//per reg: inicator beskrivelse, query-type: dg/ind
Router.get("/:register/descriptions", descriptionController);

//indicator veridier per enhet(en, flere eller ingen), unit level, per år, enheter (alle per år),
Router.get("/:register/indicators", indicatorsContoller);

//reg all-all eller per register
Router.get("/:register/unitNames", unitNamesContoller);

Router.get("/:register/years", selectionYearsContoller);

export default Router;
