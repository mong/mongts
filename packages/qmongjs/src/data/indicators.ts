export const commonIndicators = [
  "hjerteinfarkt_reper_stemi_tid",
  "hjerneslag_beh_tromb",
  "hjerneslag_rapport_3mnd",
  "NDV_andel_HbA1C_stoerre_eller_lik_75",
  "hoftebrudd_ventetid48",
  "hoftebrudd_stammefiks",
];

export const indicatorsPerHospital = [
  {
    unit: "Kirkenes",
    commonInd: commonIndicators,
    specificInd: [
      "hjerneslag_test_svelg",
      "hjerteinfarkt_invasivt_nstemi",
      "nyre_transplant_bt",
    ],
  },
  { unit: "Hammerfest", commonInd: commonIndicators, specificInd: [] },
  { unit: "Tromsø", commonInd: commonIndicators, specificInd: [] },
  {
    unit: "Harstad",
    commonInd: commonIndicators,
    specificInd: [
      "nger_kompl_intra_hys",
      "norgast_lekkasje_tykktarm",
      "norgast_saarruptur",
    ],
  },
  { unit: "Narvik", commonInd: commonIndicators, specificInd: [] },
  {
    unit: "Vesterålen",
    commonInd: commonIndicators,
    specificInd: ["hjerneslag_test_svelg", "hjerteinfarkt_ef_maalt"],
  },
  {
    unit: "Bodø",
    commonInd: commonIndicators,
    specificInd: [
      "NDV_andel_oppnaadd_kolesterolmaal_med_hjertekarsykdom",
      "norgast_lekkasje_tykktarm",
      "nger_kompl_intra_hys",
    ],
  },
  {
    unit: "Lofoten",
    commonInd: commonIndicators,
    specificInd: ["hjerteinfarkt_ef_maalt"],
  },
  { unit: "Mo i Rana", commonInd: commonIndicators, specificInd: [] },
  {
    unit: "Mosjøen",
    commonInd: commonIndicators,
    specificInd: ["NDV_andel_HbA1C_mindre_eller_lik_53"],
  },
  {
    unit: "Sandnessjøen",
    commonInd: commonIndicators,
    specificInd: ["hjerneslag_tverrfaglig"],
  },
];

export const indicatorInfo = [
  {
    indId: "hjerteinfarkt_reper_stemi_tid",
    title: "Reperfusjonsbehandling innen anbefalt tid ved STEMI",
    include: 1,
    sformat: ",.0%",
    level_green: 0.85,
    level_yellow: 0.7,
    level_direction: 1,
    registry: "Hjerteinfarkt",
  },
  {
    indId: "hjerneslag_beh_tromb",
    title: "Andel behandlet med trombolyse",
    include: 1,
    sformat: ",.0%",
    level_green: 0.195,
    level_yellow: 0.13,
    level_direction: 1,
    registry: "Hjerneslag",
  },
  {
    indId: "hjerneslag_rapport_3mnd",
    title: "Andel med rapportert status 3 måneder etter hjerneslaget",
    include: 0,
    sformat: ",.0%",
    level_green: 0.8,
    level_yellow: 0.7,
    level_direction: 1,
    registry: "Hjerneslag",
  },
  {
    indId: "NDV_andel_HbA1C_stoerre_eller_lik_75",
    title: "Andel med HbA1C ≥ 75 mmol/mol",
    include: 1,
    sformat: ",.0%",
    level_green: 0.12,
    level_yellow: 0.14,
    level_direction: 0,
    registry: "Voksendiabetes",
  },
  {
    indId: "hoftebrudd_ventetid48",
    title: "Operasjon innen 48 timer",
    include: 1,
    sformat: ",.0%",
    level_green: 0.9,
    level_yellow: 0.8,
    level_direction: 1,
    registry: "Hoftebrudd",
  },
  {
    indId: "hoftebrudd_stammefiks",
    title: "Sementert stamme ved bruk av protese",
    include: 1,
    sformat: ",.0%",
    level_green: 0.9,
    level_yellow: 0.8,
    level_direction: 1,
    registry: "Hoftebrudd",
  },
  {
    indId: "hjerneslag_test_svelg",
    title: "Andel som har fått vurdert/testet svelgefunksjon",
    include: 1,
    sformat: ",.0%",
    level_green: 0.95,
    level_yellow: 0.845,
    level_direction: 1,
    registry: "Hjerneslag",
  },
  {
    indId: "hjerteinfarkt_invasivt_nstemi",
    title: "	Koronar angiografi ved NSTEMI",
    include: 1,
    sformat: ",.0%",
    level_green: 0.85,
    level_yellow: 0.7,
    level_direction: 1,
    registry: "Hjerteinfarkt",
  },
  {
    indId: "nyre_transplant_bt",
    title: "Blodtrykk <130/80 (transplantasjon)",
    include: 1,
    sformat: ",.0%",
    level_green: 0.8,
    level_yellow: 0.7,
    level_direction: 1,
    registry: "Nyre",
  },
  {
    indId: "nger_kompl_intra_hys",
    title: "Hysteroskopi: Komplikasjoner under operasjonen",
    include: 1,
    sformat: ",.1%",
    level_green: 0.02,
    level_yellow: 0.04,
    level_direction: 0,
    registry: "Gynekologisk endoskopi",
  },
  {
    indId: "norgast_lekkasje_tykktarm",
    title:
      "Reseksjoner av tykktarm (kolon): Lekkasje fra skjøt (anastomoselekkasje)",
    sformat: ",.1%",
    include: 1,
    level_green: 0.04,
    level_yellow: 0.06,
    level_direction: 0,
    registry: "Gastrokirurgi",
  },
  {
    indId: "norgast_saarruptur",
    title: "Sårruptur",
    include: 1,
    sformat: ",.1%",
    level_green: 0.03,
    level_yellow: 0.04,
    level_direction: 0,
    registry: "Gastrokirurgi",
  },
  {
    indId: "hjerteinfarkt_ef_maalt",
    title: "Ejeksjonsfraksjon (EF) målt",
    include: 1,
    sformat: ",.0%",
    level_green: 0.8,
    level_yellow: 0.6,
    level_direction: 1,
    registry: "Hjerteinfarkt",
  },
  {
    indId: "NDV_andel_oppnaadd_kolesterolmaal_med_hjertekarsykdom",
    title:
      "Andel pasienter med kjent hjerte/karsykdom som har oppnådd kolesterolmål, LDL mindre enn 1,8 mmol/L ",
    include: 1,
    sformat: ",.0%",
    level_green: 0.6,
    level_yellow: 0.45,
    level_direction: 1,
    registry: "Voksendiabetes",
  },
  {
    indId: "NDV_andel_HbA1C_mindre_eller_lik_53",
    title: "Andel med HbA1C ≤ 53 mmol/mol",
    include: 1,
    sformat: ",.0%",
    level_green: 0.32,
    level_yellow: 0.29,
    level_direction: 1,
    registry: "Voksendiabetes",
  },
  {
    indId: "hjerneslag_tverrfaglig",
    title:
      "Andel pasienter med tverrfaglig funksjonsvurdering innen anbefalt tid",
    include: 1,
    sformat: ",.0%",
    level_green: 0.8,
    level_yellow: 0.7,
    level_direction: 1,
    registry: "Hjerneslag",
  },
];
