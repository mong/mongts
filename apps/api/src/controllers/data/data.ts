import { RequestHandler } from "express";
import { nestedDataModel } from "../../models/data";
import { parseQuery } from "./indicators";

export const dataController: RequestHandler = async (req, res) => {
  const query = parseQuery(req);
  try {
    const rows = await nestedDataModel(query.filter);
    res.json(rows);
  } catch (error) {
    const error_message =
      error instanceof Error ? error.message : String(error);
    res.status(500).json({ message: error_message });
  }
};
