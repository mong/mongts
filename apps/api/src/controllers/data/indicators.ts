import { RequestHandler, Request } from "express";
import { indicatorsModel, Filter } from "../../models/data";

export const indicatorsContoller: RequestHandler = async (req, res) => {
  const query = parseQuery(req);

  try {
    const rows = await indicatorsModel(query.filter);
    res.json(rows);
  } catch (error) {
    const error_message =
      error instanceof Error ? error.message : String(error);
    res.status(500).json({ message: error_message });
  }
};

interface Query {
  filter?: Filter;
}

function parseQuery(req: Request): Query {
  const query: Query = {};

  if (typeof req.query === "object" && !Array.isArray(req.query)) {
    query.filter = {};

    if (typeof req.query.unit_level === "string") {
      query.filter.unit_level = req.query.unit_level;
    }
    if (Array.isArray(req.query.unit_name)) {
      const unit_name = req.query.unit_name as string[];
      query.filter.unit_name = unit_name;
    }
    if (typeof req.query.year === "string") {
      const year = Number(req.query.year) || undefined;
      query.filter.year = year;
    }
    if (typeof req.query.context === "string") {
      query.filter.context = req.query.context;
    }
    if (typeof req.query.type === "string") {
      query.filter.type = req.query.type;
    }
    query.filter.register = req.params.register;
  }
  return query;
}
