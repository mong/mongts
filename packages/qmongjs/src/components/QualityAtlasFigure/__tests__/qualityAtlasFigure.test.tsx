import { render } from "@testing-library/react";
import { QualityAtlasFigure } from "..";
import { vi, describe, test, expect } from "vitest";
import { indicatorInfo } from "../../../data/indicators";
import { testdata } from "./testdata";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

const queryClient = new QueryClient();

vi.mock("next/font/google", () => ({
  Plus_Jakarta_Sans: () => ({
    style: {
      fontFamily: "mocked",
    },
  }),
}));

vi.mock("qmongjs/useIndicatorQuery", () => {
  return new Promise(() => {
    return testdata;
  });
});

describe("QualityAtlasFigure", () => {
  test("renders", () => {
    const allIndIDs = indicatorInfo.map((row) => row.indId);
    const selectedTreatmentUnits = ["BodÃ¸", "Harstad"];

    const indNameKey = indicatorInfo.map((row) => {
      return {
        indID: row.indId,
        indTitle: row.title,
        registryShortName: row.registry,
      };
    });

    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <QualityAtlasFigure
          width={1000}
          minBoxWidth={30}
          maxBoxWidth={75}
          gap={2}
          context={"caregiver"}
          year={2023}
          indIDs={allIndIDs}
          unitNames={selectedTreatmentUnits}
          indNameKey={indNameKey}
        />
        ,
      </QueryClientProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
