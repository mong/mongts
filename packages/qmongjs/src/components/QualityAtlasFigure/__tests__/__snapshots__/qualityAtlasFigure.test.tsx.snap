// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QualityAtlasFigure > renders 1`] = `
<div>
  ,
</div>
`;
